### PR TITLE
Add InterestedResources as a pod filter based on pods' requested resources to extender config

### PR DESCRIPTION
**Pull request by chenchun**
The scheduler sends filter/priority requests to the extender only if the pod to be schedulered is interested by the extender. This is helpful to scheduler performance if the extender targets on a small group of pods.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
For #53616
Fixes # https://github.com/kubernetes/kubernetes/issues/58850

**Special notes for your reviewer**:

**Release note**:
```release-note
Add InterestedResources as a pod filter based on pods' requested resources to extender config. If it's empty, every pod will be sent to extender for filter/prioritize/bind operations if FilterVerb/PrioritizeVerb/BindVerb isn't empty. Otherwise, only pods that requests for one of the resources will be.
```

### CODE REVIEW

**Comment by bsalamat** [comment]
/ok-to-test

cc: @vishh @yguo0905 Please take a look and comment if this PR satisfies the requirements that you had in mind for custom resource scheduling.

**Comment by yguo0905** [comment]
I think this label selection mechanism is a general solution, where a webhook can be used to set the proper labels on a pod based on the resources the pod requests. Ideally, it would be great if we have a way to select pod directly based on the requesting custom resources.

File: pkg/scheduler/api/types.go [hunk 1]
```diff
@@ -157,6 +157,9 @@ type ExtenderConfig struct {
 	// so the scheduler should only send minimal information about the eligible nodes
 	// assuming that the extender already cached full details of all nodes in the cluster
 	NodeCacheCapable bool
+	// Only pods with label keys and values matching this selector will be sent to the extender.
+	// +optional
+	Selector map[string]string
```
Comment (yguo0905) [inline pkg/scheduler/api/types.go]: Use metav1.LabelSelector?

File: pkg/scheduler/api/v1/types.go [hunk 2]
```diff
@@ -149,6 +149,9 @@ type ExtenderConfig struct {
 	// so the scheduler should only send minimal information about the eligible nodes
 	// assuming that the extender already cached full details of all nodes in the cluster
 	NodeCacheCapable bool `json:"nodeCacheCapable,omitempty"`
+	// Only pods with label keys and values matching this selector will be sent to the extender.
+	// +optional
+	Selector map[string]string
```
Comment (bsalamat) [inline pkg/scheduler/api/v1/types.go]: Why not using a `LabelSelector` instead of a map of key-value pairs?

**Comment by bsalamat** [comment]
> I think this label selection mechanism is a general solution, where a webhook can be used to set the proper labels on a pod based on the resources the pod requests. Ideally, it would be great if we have a way to select pod directly based on the requesting custom resources.

I see your point, but where should that filtering be implemented? I don't think we have a clean way of performing such filteration in K8s.

**Comment by ravisantoshgudimetla** [comment]
IIUC, what we are trying to do here is scheduling for custom resources using extenders.

Just thinking out loud here. Can we have a field in pod spec whose value is an extender name(or endpoint) similar to custom scheduler? Would that be a problem in long run.

File: pkg/scheduler/core/extender.go [hunk 3]
```diff
@@ -83,6 +86,16 @@ func NewHTTPExtender(config *schedulerapi.ExtenderConfig) (algorithm.SchedulerEx
 		Transport: transport,
 		Timeout:   config.HTTPTimeout,
 	}
+	if config.Selector == nil {
```
Comment (bsalamat) [inline pkg/scheduler/core/extender.go]: You should place `selector = labels.Everything()` inside this `if` and remove the other `if` below.

File: pkg/scheduler/api/types.go [hunk 4]
```diff
@@ -157,6 +157,9 @@ type ExtenderConfig struct {
 	// so the scheduler should only send minimal information about the eligible nodes
 	// assuming that the extender already cached full details of all nodes in the cluster
 	NodeCacheCapable bool
+	// Only pods with label keys and values matching this selector will be sent to the extender.
```
Comment (bsalamat) [inline pkg/scheduler/api/types.go]: You should add to the comment that this selector is applied only to filter and prioritize operations. If the extender is a binder, all the pods are sent to it for binding. Please add the comment to v1/types.go as well.

**Comment by yguo0905** [comment]
> I see your point, but where should that filtering be implemented? I don't think we have a clean way of performing such filteration in K8s.

In the scheduler? The scheduler understands the resource names, and it also knows the mapping between resource names and extenders.

**Comment by yguo0905** [comment]
> Just thinking out loud here. Can we have a field in pod spec whose value is an extender name(or endpoint) similar to custom scheduler? Would that be a problem in long run.

I'm not sure it's a good idea to expose the name of the extenders to users.

**Comment by bsalamat** [comment]
> In the scheduler? The scheduler understands the resource names, and it also knows the mapping between resource names and extenders.

That approach wouldn't be scalable. What if a user does not need to filter based on resources? We can't let everyone add an arbitrary filter to scheduler. The current approach of having a label selector is better. That way users can have an admission controller or initializer to add custom labels to pods that should be sent to a particular extender.

File: pkg/scheduler/core/extender.go (hunk 3 above)
Comment (chenchun) [reply pkg/scheduler/core/extender.go]: Updated

```
selector := labels.Everything()
if config.Selector != nil {
	selector, err = metav1.LabelSelectorAsSelector(config.Selector)
	if err != nil {
		return nil, err
	}
}
```

File: build/root/BUILD.root [hunk 5]
```diff
@@ -80,3 +80,16 @@ genrule(
     cmd = "grep ^STABLE_BUILD_SCM_REVISION bazel-out/stable-status.txt | awk '{print $$2}' >$@",
     stamp = 1,
 )
+
+go_library(
+    name = "go_default_library",
+    srcs = ["test.go"],
```
Comment (chenchun) [inline build/root/BUILD.root]: Ops, I have a temp test.go in the root directory and it makes this unrelated code gen. Will update code gen

**Comment by wanghaoran1988** [comment]
> The scheduler sends filter/priority requests to the extender only if the pod to be schedulered is interested by the extender.
Instead of saying the pod is interested by the extender, I'd like say we need schedule pods that match the filter using the extender, agree with @bsalamat , the scheduler shouldn't know what should be scheduled based on the filer, currently the scheduler only schedule the pods that have `schedulerName` eaqul to itself.
In my mind, the best approach for you should be a custom scheduler + a initializer/admission controller

@bsalamat I found more and more customer are try to using their custom scheduler, and I think we should have a scheduler_admission controller now, WDYT?

**Comment by bsalamat** [comment]
> I found more and more customer are try to using their custom scheduler, and I think we should have a scheduler_admission controller now, WDYT?

Could you elaborate a bit more on what is a scheduler_admission controller? Is it a special kind of admission controller? Basically, I would like to know what it can acheive that a generic admission contorller that a user may create cannot acheive.

**Comment by bsalamat** [comment]
/sig scheduling

LGTM
@yguo0905 do you have any other comments?

**Comment by yguo0905** [comment]
@chenchun, we had discussions offline.

- This label selector mechanism is a general solution. It requires a mutating webhook to examine all pods and set the labels accordingly, which may not scale well. Can you please describe your use case with some details?

- Also, this label selector should apply to filter, prioritize and also bind. It should either skip the extender or not, but not part of it. And we'd need to do this as a whole in order to avoid any user facing behavior change later.

**Comment by chenchun** [comment]
I'm having lots of pods using overlay network and a little of them using underlay network.
For underlay network, there is not enough ips for each rack. So I implement a scheduler extender to limit the number of pods per rack which uses underlay network. And this PR  is helpful to scheduler performance because the extender only targets on this small group of pods.
I agree with `this label selector should apply to filter, prioritize and also bind.` I'll do refactor.

**Comment by chenchun** [comment]
> It requires a mutating webhook to examine all pods and set the labels accordingly, which may not scale well.

In my case, if I'd like pods to run in overlay or underlay network, I add `network=overlay` or `network=underlay` label to pods before submitting them to k8s.

**Comment by wanghaoran1988** [comment]
> Could you elaborate a bit more on what is a scheduler_admission controller? Is it a special kind of admission controller? Basically, I would like to know what it can acheive that a generic admission contorller that a user may create cannot acheive.

IIRC, the extender is designed part of the "default-scheduler" with extra predicates and priority function, in my mind all the priority/predicates functions should works on the same group of pods, instead of some of the priority/predicates funcs interested on the small set of the group pods.

Another problems is the extender using the label selector to selec pods seems cross all the namespaces now, if we have a user case later that want to filter pods in specfic namespace, we need also add a namespace selector to let the extender interested.
For simplify,  in my mind, we need filter pods one time, if the pods have a schedulerName set, means the scheduler should responsible for scheduler this pod, the extender is part of the scheduler, should also responsible for this pod.

For the scheduler_admission controller, it's nothing special, we can make it possible to config what pods can be set to what scheduler.

**Comment by chenchun** [comment]
> For the scheduler_admission controller, it's nothing special, we can make it possible to config what pods can be set to what scheduler.

Doesn't `PodSpec.SchedulerName` is exactly what you are saying ? If a custom scheduler which is likely to be launched and maintained by the administrators rather than by the users of the cluster only watches and schedules pods pointing to it, I don't think there is a need for a controller.

> the extender is designed part of the "default-scheduler" with extra predicates and priority function, in my mind all the priority/predicates functions should works on the same group of pods, instead of some of the priority/predicates funcs interested on the small set of the group pods.

Please read my use case. The point is to help improve scheduler performance if the extender targets on a small group of pods.

**Comment by wanghaoran1988** [comment]
> Please read my use case. The point is to help improve scheduler performance if the extender targets on a small group of pods.

Why not setup a custom schduler for those specific group of pods?

**Comment by chenchun** [comment]
@wanghaoran1988 Yes, a custom scheduler can cover my case, but it definitely takes more efforts to maintain another scheduler. Scheduler extender is a simpler solution for me.

**Comment by bsalamat** [comment]
@wanghaoran1988 
> IIRC, the extender is designed part of the "default-scheduler" with extra predicates and priority function, in my mind all the priority/predicates functions should works on the same group of pods, instead of some of the priority/predicates funcs interested on the small set of the group pods.

There are reasons, particularly performance, that people may not want to run all predicate and priority functions on all pods. So, I think it makes sense to have filters.

> Another problems is the extender using the label selector to selec pods seems cross all the namespaces now, if we have a user case later that want to filter pods in specfic namespace, we need also add a namespace selector to let the extender interested.

I agree that the current filteration mechanism may not cover all cases, but disallowing filters is not the solution to this problem.

> For the scheduler_admission controller, it's nothing special, we can make it possible to config what pods can be set to what scheduler.

True, but most of our users do not want to deal with maintaining a custom scheduler. So, having multiple schedulers won't work for them.

**Comment by wanghaoran1988** [comment]
> There are reasons, particularly performance, that people may not want to run all predicate and priority functions on all pods. So, I think it makes sense to have filters.

If so, will we support the filters for the predicates/priority funcs in the default-scheduler ?  it will makes scheduler more complicated, multiple scheduler should be the best approach.

> I agree that the current filteration mechanism may not cover all cases, but disallowing filters is not the solution to this problem.

If so, we need restruct the filter maybe, and make is more extensionable.

File: pkg/scheduler/api/v1/types.go [hunk 6]
```diff
@@ -149,6 +149,9 @@ type ExtenderConfig struct {
 	// so the scheduler should only send minimal information about the eligible nodes
 	// assuming that the extender already cached full details of all nodes in the cluster
 	NodeCacheCapable bool `json:"nodeCacheCapable,omitempty"`
+	// Only pods matching this selector will be sent to the extender for filter/prioritize/bind operations.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
```
Comment (wanghaoran1988) [inline pkg/scheduler/api/v1/types.go]: If this is used to select pods, how about rename to PodSelector?

File: pkg/scheduler/api/types.go [hunk 7]
```diff
@@ -157,6 +157,9 @@ type ExtenderConfig struct {
 	// so the scheduler should only send minimal information about the eligible nodes
 	// assuming that the extender already cached full details of all nodes in the cluster
 	NodeCacheCapable bool
+	// Only pods matching this selector will be sent to the extender for filter/prioritize/bind operations.
+	// +optional
+	Selector *metav1.LabelSelector
```
Comment (yguo0905) [inline pkg/scheduler/api/types.go]: Maybe mention the default value, which is an empty selector that matches every pod.

File: pkg/scheduler/api/v1/types.go (hunk 6 above)
Comment (yguo0905) [inline pkg/scheduler/api/v1/types.go]: and here

File: pkg/scheduler/factory/factory.go [hunk 8]
```diff
@@ -928,14 +928,22 @@ func (f *configFactory) CreateFromConfig(policy schedulerapi.Policy) (*scheduler
 	return f.CreateFromKeys(predicateKeys, priorityKeys, extenders)
 }
 
-// getBinder returns an extender that supports bind or a default binder.
-func (f *configFactory) getBinder(extenders []algorithm.SchedulerExtender) scheduler.Binder {
+// getBinderFunc returns an func which returns an extender that supports bind or a default binder based on the given pod.
+func (f *configFactory) getBinderFunc(extenders []algorithm.SchedulerExtender) func(pod *v1.Pod) scheduler.Binder {
```
Comment (yguo0905) [inline pkg/scheduler/factory/factory.go]: This function becomes non-trivial now. Do we have test coverage?

**Comment by vishh** [comment]
> I'm having lots of pods using overlay network and a little of them using underlay network.
For underlay network, there are not enough ips for pods in each rack. So I implement a scheduler extender to limit the number of pods per rack which uses underlay network. And this PR is helpful to scheduler performance because the extender only targets on this small group of pods.

@jiayingz here's another use case for cluster level resources with specific topology properties (rack in this case)

> In my case, if I'd like pods to run in overlay or underlay network, I add network=overlay or network=underlay label to pods before submitting them to k8s. So I don't need a mutating webhook.

The kubernetes model does not recommend using labels to convey requirements to system components (extender in this case). Why not use annotations instead?

Here's another model to consider. Assuming `overlay` is the default choice, you can have pods request a compute resource called `hello.com/underlay-network: 1` and have the scheduler only route pods that request this resource to the corresponding extender. In the future, once we add support for cluster level counted resources, you could be able to export the number of IPs available per-rack to k8s and not use the extender at all, and not have to change your Pod Specs too.

Having ability to selectively involve the extender in the scheduling pipeline for specific resources will benefit other cluster level resource use cases too and doesn't require webhooks for example to make it a generic solution.

Thoughts? @chenchun @bsalamat @yguo0905

**Comment by chenchun** [comment]
> The kubernetes model does not recommend using labels to convey requirements to system components (extender in this case). Why not use annotations instead?

I agree, using labels to express network type is not a k8s recommend way to do it. But labels are more powerful than annotations, you can select pods by labels like what's been done in this PR.

> Here's another model to consider. Assuming overlay is the default choice, you can have pods request a compute resource called hello.com/underlay-network: 1 and have the scheduler only route pods that request this resource to the corresponding extender. In the future, once we add support for cluster level counted resources, you could be able to export the number of IPs available per-rack to k8s and not use the extender at all, and not have to change your Pod Specs too.

Is there a more detailed doc talking about this? If cluster level counted resources is implemented, I think it can replace my current scheduler extender which does the same thing of limiting numbers of pods per rack. But it may have its limitations, we have a requirement of making uderlay ip which belong to a pod of a stateful set consistent even the pod is relaunched on another node.

File: pkg/scheduler/factory/factory.go (hunk 8 above)
Comment (chenchun) [reply pkg/scheduler/factory/factory.go]: Added `TestGetBinderFunc`

**Comment by bsalamat** [comment]
I generally agree that label selectors are better than annotation filtering, at the same time, I would like this feature to have a real and reasonable use-case. If users have to create and add an admission controller to automatically add the labels based on Pods' requested resources, they may never use this feature. In such a case, I'd agree with what @vishh and @yguo0905 said that it might be better to create the filter based on requested resources.

**Comment by vishh** [comment]
@chenchun

> Is there a more detailed doc talking about this? If cluster level counted resources is implemented, I think it can replace my current scheduler extender which does the same thing of limiting numbers of pods per rack. But it may have its limitations, we have a requirement of making uderlay ip which belong to a pod of a stateful set consistent even the pod is relaunched on another node.

Cluster level resources are still in the design phase. Look forward to design docs on that in sig-scheduling. How about the other suggestion I mentioned earlier (using a compute resource now) which should work today and (possibly) meet your constraints as well?

**Comment by chenchun** [comment]
@vishh I understand your point now. I used to think that only those defined in pkg/apis/core/types.go like cpu, memory are valid computer resources and custom resources is not allowed. I'll update this PR to filter on computer resources.

**Comment by yguo0905** [comment]
@chenchun, can you please update the title to reflect the current design?

There's one more thing to note, but not for this PR. kube-scheduler currently doesn't schedule pods requesting cluster level resource like `hello.com/underlay-network: 1` because such resource is not advertised by nodes and the resource fits predicate will fail. Please see the discussion at https://github.com/kubernetes/kubernetes/issues/53616#issuecomment-364581866.

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 9]
```diff
@@ -90,11 +92,12 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
...
-		podBestEffort:             isPodBestEffort(pod),
-		podRequest:                GetResourceRequest(pod),
-		podPorts:                  schedutil.GetContainerPorts(pod),
-		matchingAntiAffinityTerms: matchingTerms,
+		pod:                             pod,
+		podBestEffort:                   isPodBestEffort(pod),
+		podRequest:                      GetResourceRequest(pod),
+		podPorts:                        schedutil.GetContainerPorts(pod),
+		matchingAntiAffinityTerms:       matchingTerms,
+		scalaResourcesHandleByExtenders: sets.NewString(),
```
Comment (vishh) [inline pkg/scheduler/algorithm/predicates/metadata.go]: nit: `scalar` here and everwhere else

File: pkg/scheduler/api/types.go [hunk 10]
```diff
@@ -170,6 +170,11 @@ type ExtenderConfig struct {
 	// so the scheduler should only send minimal information about the eligible nodes
 	// assuming that the extender already cached full details of all nodes in the cluster
 	NodeCacheCapable bool
+	// InterestedResources is a pod filter based on pods' requested resources.
+	// If it's empty, every pod will be sent to extender for filter/prioritize/bind operations if FilterVerb/PrioritizeVerb/BindVerb isn't empty.
+	// Otherwise, only pods that requests for one of the resources will be.
+	// +optional
+	InterestedResources []string
```
Comment (vishh) [inline pkg/scheduler/api/types.go]: I'd recommend setting this field as follows:
```go
type InterestedResource struct {
 // Has to pass API validation for a compute resource 
 Name v1.ResourceName
 // Scheduler avoids expecting cluster level resources to be attached to nodes
 IsClusterLevel bool
}

// This field is optional
InterestedResources []InterestedResource `json:"interestedResource, omitempty"` 
```

File: pkg/scheduler/api/validation/validation_test.go [hunk 11]
```diff
@@ -69,6 +69,14 @@ func TestValidatePolicy(t *testing.T) {
 				}},
 			expected: errors.New("Only one extender can implement bind, found 2"),
 		},
+		{
+			policy: api.Policy{
```
Comment (vishh) [inline pkg/scheduler/api/validation/validation_test.go]: Test invalid resource names.

**Comment by vishh** [comment]
Thanks for working on this @chenchun! @bsalamat I think this PR meets all of my design expectations as explained in https://github.com/kubernetes/kubernetes/issues/53616#issuecomment-365383862

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 12]
```diff
@@ -713,8 +714,12 @@ func PodFitsResources(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *s
 	}
 
 	var podRequest *schedulercache.Resource
+	scalaResourcesHandleByExtenders := sets.NewString()
 	if predicateMeta, ok := meta.(*predicateMetadata); ok {
 		podRequest = predicateMeta.podRequest
+		if predicateMeta.scalaResourcesHandleByExtenders != nil {
```
Comment (vishh) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Isn't `meta` an implementation of [this interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/types.go#L165)? How are you accessing fields directly?

Comment (vishh) [reply pkg/scheduler/algorithm/predicates/predicates.go]: nvm. I see we dynamically cast which isn't ideal though. @bsalamat FYI!

File: pkg/scheduler/factory/factory.go [hunk 13]
```diff
@@ -919,14 +919,17 @@ func (c *configFactory) CreateFromConfig(policy schedulerapi.Policy) (*scheduler
...
 		for ii := range policy.ExtenderConfigs {
 			glog.V(2).Infof("Creating extender with config %+v", policy.ExtenderConfigs[ii])
 			extender, err := core.NewHTTPExtender(&policy.ExtenderConfigs[ii])
 			if err != nil {
 				return nil, err
 			}
 			extenders = append(extenders, extender)
+			interestedResources.Insert(policy.ExtenderConfigs[ii].InterestedResources...)
 		}
+		predicates.RegisterPodFitsResourcesPredicateMetadataProducer(interestedResources)
```
Comment (vishh) [inline pkg/scheduler/factory/factory.go]: This feels odd. Why not extend the metadataproducer interface?

Comment (chenchun) [reply pkg/scheduler/factory/factory.go]: Do you mean ?

```
-type PredicateMetadataProducer func(pod *v1.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo) PredicateMetadata
+type PredicateMetadataProducer func(pod *v1.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, scalarResourcesHandleByExtenders sets.String) PredicateMetadata
```

**Comment by yguo0905** [comment]
@chenchun, code freeze for 1.10 is next Monday (2/26). Could we get this into 1.10?

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 14]
```diff
@@ -53,6 +54,7 @@ type predicateMetadata struct {
 	serviceAffinityInUse               bool
 	serviceAffinityMatchingPodList     []*v1.Pod
 	serviceAffinityMatchingPodServices []*v1.Service
+	scalaResourcesHandleByExtenders    sets.String
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/metadata.go]: Please add a comment for this. Also mention that this is set based on the policy config and does not change per pod.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 15]
```diff
@@ -1592,3 +1600,10 @@ func (c *VolumeBindingChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMe
 	glog.V(5).Infof("All PVCs found matches for pod %v/%v, node %q", pod.Namespace, pod.Name, node.Name)
 	return true, nil, nil
 }
+
+// RegisterPodFitsResourcesPredicateMetadataProducer registers a PredicateMetadataProducer used by PodFitsResources
+func RegisterPodFitsResourcesPredicateMetadataProducer(scalaResourcesHandleByExtenders sets.String) {
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Not sure if we really need this function. Please consider calling `RegisterPredicateMetadataProducer` directly.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 16]
```diff
@@ -1592,3 +1600,10 @@ func (c *VolumeBindingChecker) predicate(pod *v1.Pod, meta algorithm.PredicateMe
...
+	RegisterPredicateMetadataProducer("PodFitsResources", func(pm *predicateMetadata) {
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Please use the `const` name of the predicate instead of the string literal ("PodFitsResources").

File: pkg/scheduler/api/types.go (hunk 10 above)
Comment (yguo0905) [reply pkg/scheduler/api/types.go]: IIUC, the `IsClusterLevel` and the additional flag @bsalamat suggested are the same - they both indicate whether this resource should be skipped in the `PodFitsResources` predicate.

**Comment by yguo0905** [comment]
@chenchun, if you don't have time on this, I'm happy to take over, in order to catch the deadline next Monday. :)

File: pkg/scheduler/api/types.go (hunk 10 above)
Comment (bsalamat) [reply pkg/scheduler/api/types.go]: >  the IsClusterLevel and the additional flag @bsalamat suggested are the same

While that's true in many cases, it is not always true. A resource may be node-level, but we may want it to be ignored by the scheduler and be sent to the extender. For now, we probably want all cluster resources to be ignored though. So, if `IsClusterLevel` is true, scheduler should ignore, but if it is false, it should check the other flag that I suggested.

**Comment by chenchun** [comment]
@yguo0905 Just back to work, I'll update PR shortly.

File: pkg/scheduler/api/types.go (hunk 10 above)
Comment (chenchun) [reply pkg/scheduler/api/types.go]: > A resource may be node-level, but we may want it to be ignored by the scheduler and be sent to the extender

@bsalamat as @vishh suggested,  `IsClusterLevel` can be set to true to meet this requirement. If I'm adding another field for example `IsNodeLevel` as you suggested, scheduler does the same thing for both of them. So I think it is redundant to adding `IsNodeLevel`.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 15 above)
Comment (chenchun) [reply pkg/scheduler/algorithm/predicates/predicates.go]: `predicateMetadata` is an un-exported type and I can't calling RegisterPredicateMetadataProducer directly. Should I make it exported ?

**Comment by chenchun** [comment]
@ygu0905 Please take over on this, I may not have time to update.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 15 above)
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Oh, I see. This is fine. We should probable move this function to metadata.go.

File: pkg/scheduler/api/types.go (hunk 10 above)
Comment (yguo0905) [reply pkg/scheduler/api/types.go]: I think we can have only `IgnoredByScheduler` for now, which will satisfy the use cases discussed here.

**Comment by yguo0905** [comment]
> @yguo0905 Please take over on this, I may not have time to update.

OK, I'm working on this.

**Comment by yguo0905** [comment]
/close
Let's move the discussion to https://github.com/kubernetes/kubernetes/pull/60332
